Admin theme changes

evrifaessa opened the pull request:
- BoxBilling -> FOSSBilling changes
- See comments for others

BelleNottelling commented:
I think some modules may reference these functions that have been renamed. This could potentially be a breaking change

evrifaessa commented:
Another change. Dropped the UItoTop jQuery plugin from likely a decade ago in favor of the vanilla JS replacement I wrote now.

[Screencast from 2022-11-19 20-57-59.webm](https://user-images.githubusercontent.com/35808275/202864939-3a851bf1-fc74-45a5-8dae-141f896c0dce.webm)

Getting rid of our jQuery dependance one by one.

evrifaessa commented:
Another big change is moving templates to their modules' directories.

Some templates were in their related modules' `html_admin` directory, some were in the `bb-themes/admin_default/html` directory. To make the `admin_default/html` directory cleaner and get everything placed the same way, I moved them all under their own modules' directory.

Also, the default admin theme's templates now act as a fallback.
The theme engine first looks for the template in the active theme's `/html` directory. If it can't find it, it searchs the `html_admin` directory of the related module.

We've moved the templates to the module directories, now it will actually have some templates to use as a fallback.

evrifaessa commented:
> I think some modules may reference these functions that have been renamed. This could potentially be a breaking change

@BelleNottelling - I've now added a fallback for that. It still functions well like it used to, but logs a warning to the browser console when used as `boxbilling.message()`.

We should keep this for a few releases, and remove once we're confident everyone has jumped ship.

https://github.com/FOSSBilling/FOSSBilling/blob/b78e1bcbc9a2a6537e694e7d0c8106c77a52beef/src/bb-themes/admin_default/assets/js/fossbilling.js#L392-L398

evrifaessa commented:
Also added support for two message types other than "info". The dashboard should now recognize "warning" and "danger" too.

Here is how they look:

![image](https://user-images.githubusercontent.com/35808275/202923165-ebf41c95-b444-40cd-b8c3-8afaf677aa18.png)
![image](https://user-images.githubusercontent.com/35808275/202923180-827ea94e-b1de-496c-9c20-6134a69044ef.png)

I think it looks better in dark mode.

evrifaessa commented:
I'm also planning to add support for actions below the messages. But that will require changes to the API and I'll do it in another PR another day.

![image](https://user-images.githubusercontent.com/35808275/202923264-8b9f04ea-d39e-4205-987f-5195396a7e9b.png)

evrifaessa commented:
My only concern is how we will get people to delete the old templates from `/admin_default/html`.

If they don't delete them, there will be duplicate templates and the system will prefer the ones under `/admin_default/html`. This will result in people not receiving template updates in the future.

Maybe we can do an announcement in the Discord server? If we had a newsletter now, that would be very useful too.

BelleNottelling commented:
Add to the existing patch in foss-update.php and add a cleanup step to delete the files

evrifaessa commented on src/bb-themes/admin_default/assets/js/fossbilling.js:
```diff
@@ -257,6 +259,61 @@
+    element.setAttribute('aria-live', 'assertive');
+    element.setAttribute('aria-atomic', 'true');
+
+    element.innerHTML = `
+            <div class="toast-header">
+                <span class="p-2 border border-light bg-${color} rounded-circle me-2"></span>
+                <strong class="me-auto">System message</strong>
+                <button type="button" class="btn-close" data-bs-dismiss="toast" aria-label="Close"></button>
+            </div>
+            <div class="toast-body">${message}</div>
+        `;
```
This is triggered by the existing code. It's not a result of this PR.

evrifaessa commented:
> Add to the existing patch in foss-update.php and add a cleanup step to delete the files

I'm not sure if that's the best way to do it. I wouldn't want to delete people's templates with each update.

For those migrating from BB, they already will have to delete the whole old admin theme and install this instead. What I'm concerned about is the existing FOSSBilling users. I think it would be just fine if we do an announcement in the Discord server.

To avoid these in the future, we might want to start a newsletter or add a "notifications"/"news" area that pulls the news and notices from our website and displays them in the admin panel.

evrifaessa commented:
Got rid of another jQuery plugin, "tipsy".

It was used for tooltips in the old admin panel theme. [Tabler](https://preview.tabler.io/docs/tooltips.html) already has support for tooltips built in. I also couldn't find any places where Tipsy was being used.

Before the scope of this PR gets any bigger, I'd appreciate if we review and merge it. I have other changes in mind but want to do them in separate PRs.

evrifaessa commented:
Dropped jquery.collapsible too. Tabler has it built in and we weren't using it anywhere.

evrifaessa commented:
Also jquery.jeditable. It wasn't used anywhere. It wasn't even specified to be used in the Gulpfile.

evrifaessa commented:
I've checked #113 to see what we can use. Included ~10 pages rewritten in Tabler from that PR into this one. I wish we could hear from Andrey.

We are now closer to having zero pages written for the old admin theme. I assume we only have <10 or <15 pages left.

I'm going to bed. This should be ready for reviews. @BelleNottelling @John-S4.

BelleNottelling reviewed (approved):
LGTM. Tested the artifact and everything does appear to be working correctly, and these changes make sense
